### PR TITLE
Volume calculation

### PR DESCRIPTION
**Pull request by DaumantasJankauskas**
-Add queries and methods for physical elements volume calculation.
-Will update query file structure:
--Change output directory hierarchy
--Update test case
-Update how queries are called for execution.
-Update Readme.md file
--Add status badge for ci workflow

### CODE REVIEW

**Review by ColinKerr — COMMENTED** [review]
What if we used this format for the json file?
```json
    "queries" : {
      "outputName": {
        "info" : "Just for info/notes/docs about query ... not necessarily displayed or written anywhere.  Object key is the filename to use.",
        "query" : "SELECT * FROM BisCore.PhysicalElement",
        "options" : {
          "calculateMassProperties" : true,
          "idColumn" : 0,
          "idColumnIsJsonArray" : false
        },
        "schema" : {
          "query" :"SELECT DISTINCT schema.Name, schema.VersionMajor, schema.VersionWrite, schema.VersionMinor, schema.DisplayLabel, schema.Description FROM ECDbMeta.ECSchemaDef schema JOIN ECDbMeta.ECClassDef class ON class.Schema.Id = schema.ECInstanceId WHERE class.ECInstanceId in (SELECT DISTINCT(ECClassId) FROM Bis.Element)"
        },
        "class" : {
          "info" : "The query above is the bare minimum, info and options may be null calculateMassProperties defaults to false, idColumn defaults to 0 and idColumnIsJsonArray defaults to false.  idColumn gives the position of the column which holds the ids to use when calculating the mass props.",
          "query" : "SELECT COUNT(e.ECInstanceId) as [Count], e.ECClassId, class.DisplayLabel, class.Description FROM Bis.Element e JOIN ECDbMeta.ECClassDef class ON class.ECInstanceId = e.ECClassId GROUP BY e.ECClassId ORDER BY ec_classname(e.ECClassId)"                        
        }
      }
    }
```
The benefit of setting it up this way is we could parse each query with the same code then have the child method do different things based on the options.  We would overall have less code and it would be easier to add more permutations in the future without making the code more complex.

One thing that I should have mentioned at the start of all this.  I never really dug into what happens when you try to use this on 2d geometry.  Please test and see what happens.  

Maybe we should let the user specify what mass properties to calculate (Volume, Area, Length) ... I don't know if they behave differently between 2d/3d elements.

What do you guys think about the new format?

**Comment by ColinKerr** [comment]
Tests need to be added for the new functionality.  Consider that someone (plausibly me) will have to keep this sample updated and working as future iModel.js releases come out.  So tests that verify basic functionality will really help keep that update process smooth and ensures the sample will always work when people try it.

File: queries/example.json [hunk 1]
```diff
@@ -1,24 +1,39 @@
 {    
     "name": "example",
     "description": "simple example queries on how tool works",
-    "url" : "https://connect-imodelweb.bentley.com/imodeljs/?projectId=<put your project id here>&iModelId=<put your model id here>&ChangeSetId=<put your changeset id here>",
+    "url" : "https://connect-imodelweb.bentley.com/imodeljs/?projectId=84869c95-46f4-4c99-a9d0-144a267db609&iModelId=488ec3ce-eb6b-4996-8279-5069b734fae4",
```
Comment (ColinKerr) [inline queries/example.json]: ids snuck in ... oops

File: src/DataExporter.ts [hunk 2]
```diff
@@ -51,33 +57,98 @@ export class DataExporter {
...
+
+  private async writeVolume(statement: ECSqlStatement, outputFileName: string, options: Options): Promise<void> {
+    const writeStream = fs.createWriteStream(outputFileName);
+    let rowCount = 0;
+    let ids: Id64Array = [];
+    const header: string[] = ["volume","area"];
+
+    let outHeader: string;
+    if (options.idColumnIsJsonArray) {
+      header.push("CodeValue");
```
Comment (ColinKerr) [inline src/DataExporter.ts]: I don't understand this.  Why are you pushing CodeValue as the header? Why not just treat this like any other query and populate the headers from the statement.  

I know the id column will be nasty but that's OK ... it's not supposed to look nice it's supposed to be the raw data.

File: src/DataExporter.ts [hunk 3]
```diff
@@ -51,33 +57,98 @@ export class DataExporter {
...
+    }
+
+    writeStream.on("finish", () => {
+     console.log(`Written ${rowCount} rows to file: ${outputFileName}`);
     });
+
+    writeStream.end();
+  }
+
+  private async defaultWrite(statement: ECSqlStatement, outputFileName: string): Promise<void> {
```
Comment (ColinKerr) [inline src/DataExporter.ts]: I think defaultWrite and writeVolume could be one method and be simple.  The two will be pretty similar if you remove the special handling for the ids as a json array.  As I noted above we can just push all the columns to the csv file same as if it was a normal query.  The only difference is that we're adding a few columns on to the beginning of the line.

File: README.md [hunk 4]
```diff
@@ -77,33 +81,52 @@ An example query file with four simple queries.
...
-	      "2dElements" : {
-            "store" : "2dElements",
-            "query" : "SELECT element.ECClassId, element.ECInstanceId ElementId, element.UserLabel, element.CodeValue FROM bis.GeometricElement2d element"
-        }
+        "volumeQueriesForGroupIds": {
+			"groupedIds" : {
+                "store" : "groupedIds",
+                "query" : "SELECT json_group_array(IdToHex(e.ECInstanceId)) as id_list, c.codevalue FROM bis.physicalElement e JOIN bis.Category c ON e.Category.Id = c.ECInstanceId GROUP BY e.Category.Id"
+            }
+        }
```
Comment (ColinKerr) [inline README.md]: Documentation not changed to reflect new json format

**Review by ColinKerr — APPROVED** [review]
This looks good, just address the readme issues and you can merge

File: README.md [hunk 5]
```diff
@@ -68,6 +70,8 @@ Query file structure below.
 ### Example query
```
Comment (AureMonke) [inline README.md:71]: Example query/Project structure/Testing some section text are in bold, difficult to read.

File: README.md [hunk 6]
```diff
@@ -77,33 +81,52 @@ An example query file with four simple queries.
...
+            "schema": {
+                "store" : "schema",
+                "query" : "SELECT DISTINCT schema.Name, schema.VersionMajor, schema.VersionWrite, schema.VersionMinor, schema.DisplayLabel, schema.Description FROM ECDbMeta.ECSchemaDef schema JOIN ECDbMeta.ECClassDef class ON class.Schema.Id = schema.ECInstanceId WHERE class.ECInstanceId in (SELECT DISTINCT(ECClassId) FROM Bis.Element)"
+            },
+            "classes" : {
+                "store" : "class",
+                "query" : "SELECT COUNT(e.ECInstanceId) as [Count], e.ECClassId, class.DisplayLabel, class.Description FROM Bis.Element e JOIN ECDbMeta.ECClassDef class ON class.ECInstanceId = e.ECClassId GROUP BY e.ECClassId ORDER BY ec_classname(e.ECClassId)"                        
+            },
+            "3dElements" : {
+                "store" : "3dElements",
```
Comment (AureMonke) [inline README.md]: We don't use "store" anymore in the code for file storing.

File: queries/example.json [hunk 7]
```diff
@@ -3,22 +3,37 @@
...
+          "query" : "SELECT COUNT(e.ECInstanceId) as [Count], e.ECClassId, class.DisplayLabel, class.Description FROM Bis.Element e JOIN ECDbMeta.ECClassDef class ON class.ECInstanceId = e.ECClassId GROUP BY e.ECClassId ORDER BY ec_classname(e.ECClassId)"                        
         },
         "3dElements" : {
-            "store" : "3dElements",
-            "query" : "SELECT element.ECClassId, element.ECInstanceId ElementId, element.UserLabel, element.CodeValue FROM bis.GeometricElement3d element"
+          "store" : "3dElements",
+          "query" : "SELECT element.ECClassId, element.ECInstanceId ElementId, element.UserLabel, element.CodeValue FROM bis.GeometricElement3d element"
+        },
+        "2dElements" : {
+          "store" : "2dElements",
```
Comment (AureMonke) [inline queries/example.json]: If we don't use "store", let's get rid of it.

File: src/MainSnapshot.ts [hunk 8]
```diff
@@ -35,14 +35,11 @@ export async function mainSnapshot(process: NodeJS.Process): Promise<void> {
     const exporter = new DataExporter(sourceDb);
     exporter.setfolder(userdata.folder);
 
-    let queryCount = 0;
```
Comment (AureMonke) [inline src/MainSnapshot.ts:38]: You have two calls to IModelHost.startup()  (line 16, line 31). Call startup after all checks of existance file, etc.

File: src/tests/MainSnapshot.test.ts [hunk 9]
```diff
@@ -31,18 +32,51 @@ describe("DataExporter.test.ts",()=> {
...
-      }
-      const outDir = path.join(__dirname, "/../../out/" + userdata.folder);
+    for (const querykey of Object.keys(userdata.queries)) {
+      const aQuery = userdata.queries[querykey];
+      await exporter.writeQueryResultsToCsv(aQuery.query, querykey + ".csv", aQuery.options)
+    }
+   const outDir = path.join(__dirname, "/../../out/" + userdata.folder);
+
+    expect(IModelJsFs.existsSync(outDir)).to.equal(true);
+    expect(IModelJsFs.readdirSync(outDir)).to.have.members(outFiles);
```
Comment (AureMonke) [inline src/tests/MainSnapshot.test.ts:47]: We just look for an exact count of files, but we don't look what's inside of these file. May we check for file not being empty / file has predefined headers (like in volume query case) / file content, just to be sure that file can be opened and content is readable.

**Review by AureMonke — APPROVED** [review]

**Pull request merged**